**PYTHON EX 1-Madison Moye-November 13, 2019**

In [0]:
import pandas as pd

DATA_URL = 'http://128.138.93.164/aprd6342/data/TravelPonyFacebook.csv'
df = pd.read_csv(DATA_URL)

In [0]:
df.head()

Start Date  ... Cost per Page Engagement (USD)
0  2013-12-02  ...                       1.219512
1  2013-12-01  ...                       1.395349
2  2013-11-30  ...                       1.190476
3  2013-11-29  ...                       1.167315
4  2013-11-22  ...                       2.512857

[5 rows x 20 columns]

In [0]:
!pip install arrow

In [0]:
import numpy as np
import arrow as arrow

##Question 1: When considering 'cost per impression', what day of the week works best? (What day is it cheapest to generate impressions)

What day works worst? (What day is most expensive)



In [0]:
#costperimpression = amountspent/impressions
df['cost per impression']=df['Amount Spent (USD)']/df['Impressions']

In [0]:
df['Start Date']=pd.to_datetime(df['Start Date'])
df['Day of Week']=df['Start Date'].dt.day_name()
df.columns

Index(['Start Date', 'End Date', 'Account', 'Campaign Name', 'Reach',
       'Frequency', 'Impressions', 'Clicks', 'Unique Clicks',
       'Amount Spent (USD)', 'Page Likes', 'Page Engagement',
       'Post Engagement', 'Post Likes', 'Post Comments', 'Post Shares',
       'Photo Views', 'Website Clicks', 'Cost per Page Like (USD)',
       'Cost per Page Engagement (USD)', 'cost per impression', 'Day of Week'],
      dtype='object')

In [0]:
df.groupby('Day of Week', as_index=False).agg({'cost per impression':'mean'})

Day of Week  cost per impression
0      Friday             0.004097
1      Monday             0.002910
2    Saturday             0.002629
3      Sunday             0.003688
4    Thursday             0.003398
5     Tuesday             0.003040
6   Wednesday             0.002981

####To answer this problem, I ended up using the mean of cost per impression; however, I am thinking I could have just divided count(day). Anyways by using the mean, it shows that the smallest average (or cheapest day to generate impressions) would be on Saturday, with a CPI of 0.002629. 



##Question 2: Which correlation is the strongest? What does that mean practically? (respond in a tweet or less)

Compute the correlation (Links to an external site.) between Amount Spent and the following variables:
- Reach
- Frequency
- Unique Clicks
- Page Likes


In [0]:
correlation = df[['Amount Spent (USD)', 'Reach', 'Frequency', 'Unique Clicks', 'Page Likes']]
correlation.corr()

Amount Spent (USD)     Reach  ...  Unique Clicks  Page Likes
Amount Spent (USD)            1.000000  0.703124  ...       0.882993    0.757612
Reach                         0.703124  1.000000  ...       0.722249    0.304388
Frequency                     0.130201  0.334101  ...       0.135103    0.000182
Unique Clicks                 0.882993  0.722249  ...       1.000000    0.584614
Page Likes                    0.757612  0.304388  ...       0.584614    1.000000

[5 rows x 5 columns]

####"Unique Clicks" (0.88-*super* close to 1) sits as the highest correlation with Amount Spent (USD) based off of the above correlation matrix. 

##Question 3: What variable most strongly predicts unique clicks? What does that mean practically? (respond in a tweet or less)

In [0]:
import statsmodels.api as sm

Unique Clicks is the dependent variable and Reach and Frequency are the independent (predictor) variables

- dependent = y
- independent = x

In [0]:
x=df[['Reach', 'Frequency']]
y=df['Unique Clicks']

In [0]:
model=sm.OLS(y,x).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:          Unique Clicks   R-squared (uncentered):                   0.557
Model:                            OLS   Adj. R-squared (uncentered):              0.556
Method:                 Least Squares   F-statistic:                              2325.
Date:                Thu, 14 Nov 2019   Prob (F-statistic):                        0.00
Time:                        05:44:06   Log-Likelihood:                         -15973.
No. Observations:                3705   AIC:                                  3.195e+04
Df Residuals:                    3703   BIC:                                  3.196e+04
Df Model:                           2                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Reach          0.0019   3.12e-05     62.490      0.000       0.002       0.002
Frequency      3.6139      0.298     12.109      0.000       3.029       4.199
==============================================================================
Omnibus:                     5107.616   Durbin-Watson:                   0.803
Prob(Omnibus):                  0.000   Jarque-Bera (JB):          5128392.747
Skew:                           7.331   Prob(JB):                         0.00
Kurtosis:                     184.674   Cond. No.                     9.84e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 9.84e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

####After performing this regression analysis, we can see that the relationship cannot be fully explained due to the R-squared value. Therefore, we cannot determine which one is more predictive than the other only that they are both are predictive. 